### PR TITLE
MainDB v3

### PR DESCRIPTION
**Pull request by qianbin**

### CODE REVIEW

**Comment by qianbin** [comment]
* reduces the complexity of states access to close to O(1) 
* the more compact storage scheme reduces disk usage by 45% (full-archived node) 
* the more efficient trie pruning algorithm
* ...

File: api/accounts/accounts.go [hunk 1]
```diff
@@ -67,15 +66,15 @@ func (a *Accounts) handleGetCode(w http.ResponseWriter, req *http.Request) error
 	if err != nil {
 		return err
 	}
-	code, err := a.getCode(addr, h.StateRoot())
+	code, err := a.getCode(addr, h)
```
[the remark this answers is not in the document]
Comment (qianbin) [reply api/accounts/accounts.go]: updated

File: chain/chain.go [hunk 2]
```diff
@@ -255,41 +306,25 @@ func (c *Chain) FindBlockHeaderByTimestamp(ts uint64, flag int) (header *block.H
...
 
-func (r *Repository) indexBlock(parentIndexRoot thor.Bytes32, block *block.Block, receipts tx.Receipts) (thor.Bytes32, error) {
-	txs := block.Transactions()
-	if len(txs) != len(receipts) {
-		return thor.Bytes32{}, errors.New("txs count != receipts count")
-	}
-
-	trie := r.db.NewTrie(IndexTrieName, parentIndexRoot)
-	id := block.Header().ID()
+func (r *Repository) indexBlock(parentIndexRoot thor.Bytes32, parentConflicts uint32, newBlock *block.Block, newConflicts uint32) (thor.Bytes32, error) {
```
Comment (libotony) [inline chain/chain.go:317]: `*block.Header` is enough

File: logdb/logdb.go [hunk 3]
```diff
@@ -355,190 +370,221 @@ func (db *LogDB) HasBlockID(id thor.Bytes32) (bool, error) {
...
+			txOrigin, _ = tx.Origin()
+
+		}
 		if err := w.exec(
-			"INSERT OR IGNORE INTO ref(data) VALUES(?)",
-			id.Bytes()); err != nil {
+			"INSERT OR IGNORE INTO ref(data) VALUES(?),(?)",
+			txID[:], txOrigin[:]); err != nil {
 			return err
 		}
```
[the remark this answers is not in the document]
Comment (qianbin) [reply logdb/logdb.go:458]: len(txs) is 0 for genesis block

**Review by libotony — APPROVED** [review]
LGTM

**Pull request merged**